feat: Add `text-transform` attribute to the `style` block

alexstoick opened the pull request:
This provides an option to override the functionality provided by the
theme, or apply `text-transform` outside of the theme.

The functionality of the `text-transform` is as below:

- `text-transform: none` - will disable **any** transformation (like the uppercasing by `terminal` theme)
- `text-transform: uppercase` (uppercase not upper as per your message) - will force all characters into uppercase.
- `text-transform: lowercase` - will force all characters into lowercase.
- `text-transform: capitalize` - will uppercase the first letter of every word

In addition, this commit introduces:
- helper methods on the `d2graph.Style` struct to determine the type of
  `text-transform` to be applied.
- `ApplyTextTransform` method on the `d2graph.Attributes` which will
  transform the `Label.Value` to the correct text case.

## Screenshots pending.

Fixes: https://github.com/terrastruct/d2/issues/1117

<!-- Remember to update ci/release/changelogs/next.md, the manpage and cli help documentation. -->

alixander reviewed (changes requested):
amazing, thanks so much for opening a PR for this.

1. your intuition is right about the boolean type, but we should make all of them typed in a refactor and test rigorously, vs having a temporary one-off type in the list of styles. (there's an issue for this: https://github.com/terrastruct/d2/issues/285). could you use the same `Scalar` type here please?
2. Tests: https://github.com/terrastruct/d2/blob/a53eb48676772409e0f3f85e3eac130263278f48/e2etests/themes_test.go#L13

- Add a caps lock = false for an object in there
- Regenerate the test with `TESTDATA_ACCEPT=1 ./ci/test.sh ./e2etests -run TestE2E/themes`
- Check the output SVG to see that caps lock=false worked.

And then do that for another test to test capslock=true works.

alexstoick commented:
@alixander - updated as per your comment:
- removed `Boolean` class. I kept the helper methods - as I think the duplication would be mad, but if you consider it's not in line with the rest of the style I can remove them 👍 
- added tests & updated description with screenshots.

alixander commented:
@janydoe that's an interesting point. maybe we literally just do https://developer.mozilla.org/en-US/docs/Web/CSS/text-transform

yeah that's more flexible. i've found myself wanting to do capital casing sometimes, and i think many others would to. So then the way to disable the theme's uppercasing is to do `text-transform: none`

Sorry for the moving goal post @alexstoick , would you mind changing to that?

alexstoick commented:
@alixander I'm happy to implement the `text-transform` spec. Just to be clear on desired behaviour:

- `text-transform: none` - will disable **any** transformation (like the uppercasing by `terminal` theme)
- `text-transform: uppercase` (uppercase not upper as per your message) - will force all characters into uppercase.
- `text-transform: lowercase` - will force all characters into lowercase.
- `text-transform: capitalize` - will uppercase the first letter of every word (defined by some separator) - I'll need to check if there is a helper method for this, or I might come back to you with a proposal.


If the above sounds good - LMK and I'll update the PR. 

Thank you 🙇

alixander commented:
@alexstoick sounds good to me!

alexstoick commented:
@alixander - I've implemented the `text-transform` as requested; before I go ahead and add final touches to description (screenshots, etc) - let me know if this is 🆗 for you. I've changed a few of the tests to cover all of the `text-transform` options.

alixander commented on d2graph/d2graph.go:
```diff
@@ -131,6 +133,22 @@ type Attributes struct {
 }
 
+func (a *Attributes) ApplyTextTransform() {
+	if a.Style.NoTextTransform() {
+		return
+	}
+
+	if a.Style.UppercaseTextTransform() {
+		a.Label.Value = strings.ToUpper(a.Label.Value)
+	}
+	if a.Style.LowercaseTextTransform() {
+		a.Label.Value = strings.ToLower(a.Label.Value)
+	}
+	if a.Style.CapitalizeTextTransform() {
+		a.Label.Value = cases.Title(language.Und).String(a.Label.Value)
+	}
```
inline the method calls please. no need to jump around for a one-liner used in one place

alixander commented on d2graph/d2graph.go:
```diff
@@ -340,6 +375,14 @@ func (s *Style) Apply(key, value string) error {
+			break
+		}
+		if !go2.Contains(textTransforms, strings.ToLower(value)) {
+			return errors.New(`expected "text-transform" to be one of (none, uppercase, lowercase, capitalize)`)
```
use strings.join on textTransforms to emit the list here

alixander commented on d2graph/d2graph.go:
```diff
@@ -151,25 +169,42 @@ func (r Reference) InEdge() bool {
+	TextTransform *Scalar `json:"capsLock,omitempty"`
+}
+
+func (s Style) NoTextTransform() bool {
```
subtle, but this should be `NoneTextTransform`. it's a special value set, otherwise can be confused with having no value (e.g. s.TextTransform == nil)

alixander replied:
or i guess this is moot given the comment about inlining these instead

alexstoick replied:
@alixander -  I'm happy to inline _all_ of them **except this one** - as this is being called from a few places. I can rename this to `NoneTextTransform` to avoid having this `if` clause all over the code. 

Will make life much easier once we come back to refactoring these to not all use `d2graph.Scalar`.

You OK with this?

alixander replied:
oh i see. yeah that's fine

alixander commented on d2graph/style_helper.go:
```diff
@@ -0,0 +1,3 @@
+package d2graph
+
+var textTransforms = []string{"none", "uppercase", "lowercase", "capitalize"}
```
just define these here plz: https://github.com/terrastruct/d2/blob/2e9c05410ccb9d6c457206d15b55386464d2283d/d2graph/d2graph.go#L1621

alixander reviewed (changes requested):
generally looks good to me, let's add tests and get this in, ty @alexstoick !

alexstoick commented:
@alixander - when you say `let's add tests` - I have already updated some of the tests in `e2e/themes`; are there more tests that you would like to see on specific components, or new tests all together?

alixander commented:
oh, i guess those tests do cover it all, nvm!

alexstoick commented on d2graph/d2graph.go:
```diff
@@ -151,25 +169,30 @@ func (r Reference) InEdge() bool {
+	Underline     *Scalar `json:"underline,omitempty"`
+	Filled        *Scalar `json:"filled,omitempty"`
+	DoubleBorder  *Scalar `json:"doubleBorder,omitempty"`
+	TextTransform *Scalar `json:"capsLock,omitempty"`
```
eagle eyes - nice one!

alixander reviewed (commented):
can you add a changelog here too please?

alixander commented:
@alexstoick looks like a format error, if you run `./make.sh fmt` it should fix

alexstoick commented:
@alixander - yep; odd that my formatter didn't auto-fix. Updated now ✅

alixander commented on d2graph/d2graph.go:
```diff
@@ -1475,16 +1510,9 @@ func (g *Graph) SetDimensions(mtexts []*d2target.MText, ruler *textmeasure.Ruler
 func (g *Graph) Texts() []*d2target.MText {
 	var texts []*d2target.MText
 
-	capsLock := g.Theme != nil && g.Theme.SpecialRules.CapsLock
-
 	for _, obj := range g.Objects {
 		if obj.Attributes.Label.Value != "" {
 			text := obj.Text()
-			if capsLock && !strings.EqualFold(obj.Attributes.Shape.Value, d2target.ShapeCode) {
-				if obj.Attributes.Language != "latex" {
-					text.Text = strings.ToUpper(text.Text)
-				}
-			}
```
how come these are removed?

alexstoick replied:
We are altering the `obj.Attributes.Label.Value` when we call the `ApplyTextTransform` method, which in turn means we no longer need to call `ToUpper` on it here.

alixander replied:
this function is not called in the course of compilation. it's just an exposed API for getting all the texts, so they still have to be applied here.

alexstoick replied:
Ok. I thought this would be covered in a test; so when I saw all text still looked as expected I thought it was safe to RM.

I will make the required changes to keep the if clause and add the other transformations.

alixander replied:
yeah there probably should be tests for it

alixander commented on d2graph/d2graph.go:
```diff
@@ -1513,9 +1541,6 @@ func (g *Graph) Texts() []*d2target.MText {
 	for _, edge := range g.Edges {
 		if edge.Attributes.Label.Value != "" {
 			text := edge.Text()
-			if capsLock {
-				text.Text = strings.ToUpper(text.Text)
-			}
```
and this

alixander commented on d2graph/d2graph.go:
```diff
@@ -1512,8 +1548,9 @@ func (g *Graph) Texts() []*d2target.MText {
+			edge.Attributes.ApplyTextTransform()
 			text := edge.Text()
-			if capsLock {
+			if capsLock && edge.Attributes.Style.NoneTextTransform() {
```
!NoneTextTransform

alixander reviewed (approved):
tested and works great, thanks @alexstoick
